### PR TITLE
Function to validate pg_balancer is enabled or not

### PR DESCRIPTION
**Pull request by ramkiperiy**
Function to validate pg_balancer is enabled or not

Signed-off-by: Ramakrishnan <rperiyas@redhat.com>

### CODE REVIEW

**Comment by ramkiperiy** [comment]
Local run logs 
`------------------------------------------------------------------------------- live log call -------------------------------------------------------------------------------
18:06:56 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/2ndMar/auth/kubeconfig get Pod  -n openshift-storage --selector=app=rook-ceph-tools -o yaml
18:06:57 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/2ndMar/auth/kubeconfig get Pod rook-ceph-tools-7f7d86c8cc-nhcg8 -n openshift-storage
18:06:58 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/2ndMar/auth/kubeconfig rsh rook-ceph-tools-7f7d86c8cc-nhcg8 ceph balancer status --format json-pretty
18:07:02 - MainThread - root - INFO - PG balancer is active
18:07:02 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/2ndMar/auth/kubeconfig get Pod  -n openshift-storage --selector=app=rook-ceph-tools -o yaml
18:07:03 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/2ndMar/auth/kubeconfig get Pod rook-ceph-tools-7f7d86c8cc-nhcg8 -n openshift-storage
18:07:04 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/2ndMar/auth/kubeconfig rsh rook-ceph-tools-7f7d86c8cc-nhcg8 ceph osd df --format json-pretty
18:07:08 - MainThread - root - WARNING - OSD data consumption difference is 6.909314060397811% between high and low utilized OSDs
FAILED                                                                                                                                                                [100%]
----------------------------------------------------------------------------- live log teardown -----------------------------------------------------------------------------`

File: ocs_ci/ocs/cluster.py [hunk 1]
```diff
@@ -743,3 +743,42 @@ def validate_osd_utilization(osd_used=80):
...
+        logging.warn("PG balancer is not active")
+        return False
+
+    # Check OSD utilization with pg balancer is active
+    osd_dict = get_osd_utilization()
+    osd_less_uti_value = min(osd_dict.values())
+    osd_high_uti_value = max(osd_dict.values())
+    diff = osd_high_uti_value - osd_less_uti_value
+    # TODO: For now kept OSD difference as 2% but have to revisit this
+    if diff <= 2:
```
Comment (keesturam) [inline ocs_ci/ocs/cluster.py]: can we get this data from ceph QE or ceph test results?

File: ocs_ci/ocs/cluster.py [hunk 2]
```diff
@@ -743,3 +743,42 @@ def validate_osd_utilization(osd_used=80):
...
+        logging.info(
+            f"OSD data consumption difference is {diff}% "
+            f"between high and low utilized OSDs"
+        )
+        return True
+    else:
+        logging.warn(
```
Comment (keesturam) [inline ocs_ci/ocs/cluster.py]: should this be log.error since we are returning false?

File: ocs_ci/ocs/cluster.py (hunk 1 above)
Comment (ramkiperiy) [reply ocs_ci/ocs/cluster.py]: Discussed with Ceph QE below are the pointers.
1. ceph balancer status --> Should have 'active' : True & they are preferring `balancer mode`: upmap, So let me add one more check for 'upmap'
2. Its not necessary OSDs capacity or PG count will be same with respect to large ceph cluster, Planning to ask Neha or Josh to gather more information on what could be the correct validation.

File: ocs_ci/ocs/cluster.py [hunk 3]
```diff
@@ -743,3 +743,42 @@ def validate_osd_utilization(osd_used=80):
...
+    ceph_cmd = "ceph balancer status"
+    ct_pod = pod.get_ceph_tools_pod()
+    output = ct_pod.exec_ceph_cmd(ceph_cmd=ceph_cmd)
+    if output['active']:
+        logging.info("PG balancer is active")
+    else:
+        logging.warn("PG balancer is not active")
+        return False
+
+    # Check OSD utilization with pg balancer is active
```
[the remark this answers is not in the document]
Comment (ramkiperiy) [reply ocs_ci/ocs/cluster.py]: ack

File: ocs_ci/ocs/cluster.py (hunk 2 above)
Comment (ramkiperiy) [reply ocs_ci/ocs/cluster.py]: ack

File: ocs_ci/ocs/cluster.py (hunk 1 above)
Comment (ramkiperiy) [reply ocs_ci/ocs/cluster.py]: added more checks after discussing with Ceph QE

**Comment by ramkiperiy** [comment]
Local run logs for function pass and failure.
`------------------------------------------------------------------------------- live log call -------------------------------------------------------------------------------
15:30:21 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/vmware/auth/kubeconfig get Pod  -n openshift-storage --selector=app=rook-ceph-tools -o yaml
15:30:23 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/vmware/auth/kubeconfig get Pod rook-ceph-tools-644d48b77f-dxxjw -n openshift-storage
15:30:24 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/vmware/auth/kubeconfig rsh rook-ceph-tools-644d48b77f-dxxjw ceph balancer status --format json-pretty
15:30:29 - MainThread - root - INFO - Print {'plans': [], 'active': False, 'last_optimize_started': '', 'last_optimize_duration': '', 'optimize_result': '', 'mode': 'none'}
15:30:29 - MainThread - root - ERROR - PG balancer is not active
15:30:29 - MainThread - root - INFO - pg_balancer is not active
FAILED                                                                                                                                                                [100%]

================================================================================= FAILURES ==================================================================================

------------------------------------------------------------------------------- live log call -------------------------------------------------------------------------------
16:44:28 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/vmware/auth/kubeconfig get Pod  -n openshift-storage --selector=app=rook-ceph-tools -o yaml
16:44:29 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/vmware/auth/kubeconfig get Pod rook-ceph-tools-644d48b77f-dxxjw -n openshift-storage
16:44:31 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/vmware/auth/kubeconfig rsh rook-ceph-tools-644d48b77f-dxxjw ceph balancer status --format json-pretty
16:44:36 - MainThread - root - INFO - PG balancer is active and mode is upmap
16:44:36 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/vmware/auth/kubeconfig get Pod  -n openshift-storage --selector=app=rook-ceph-tools -o yaml
16:44:38 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/vmware/auth/kubeconfig get Pod rook-ceph-tools-644d48b77f-dxxjw -n openshift-storage
16:44:39 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/vmware/auth/kubeconfig rsh rook-ceph-tools-644d48b77f-dxxjw ceph osd df --format json-pretty
16:44:44 - MainThread - root - INFO - Eval value is 0.000732 and pg distribution difference is 0 between high and low pgs per OSD
PASSED                                                                                                                                                                [100%]

============================================================================ 1 passed in 29.93s =============================================================================

------------------------------------------------------------------------------- live log call -------------------------------------------------------------------------------
16:47:54 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/vmware/auth/kubeconfig get Pod  -n openshift-storage --selector=app=rook-ceph-tools -o yaml
16:47:56 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/vmware/auth/kubeconfig get Pod rook-ceph-tools-644d48b77f-dxxjw -n openshift-storage
16:47:58 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/vmware/auth/kubeconfig rsh rook-ceph-tools-644d48b77f-dxxjw ceph balancer status --format json-pretty
16:48:03 - MainThread - root - INFO - PG balancer is active and mode is upmap
16:48:03 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/vmware/auth/kubeconfig get Pod  -n openshift-storage --selector=app=rook-ceph-tools -o yaml
16:48:04 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/vmware/auth/kubeconfig get Pod rook-ceph-tools-644d48b77f-dxxjw -n openshift-storage
16:48:06 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/vmware/auth/kubeconfig rsh rook-ceph-tools-644d48b77f-dxxjw ceph osd df --format json-pretty
16:48:11 - MainThread - root - ERROR - Eval value is 0.030732 and pg distribution difference is 0 between high and low pgs per OSD
FAILED                                                                                                                                                                [100%]

================================================================================= FAILURES ==================================================================================`

**Comment by ramkiperiy** [comment]
Travis CI failure is due to https://github.com/red-hat-storage/ocs-ci/issues/1573

**Comment by ramkiperiy** [comment]
@RazTamir  @ebenahar Would request to review and approve this PR

File: ocs_ci/ocs/cluster.py [hunk 4]
```diff
@@ -743,3 +743,92 @@ def validate_osd_utilization(osd_used=80):
...
+        logging.error("PG balancer is not active")
+        return False
+
+
+def validate_pg_balancer():
+    """
+    Validate either data is equally distributed to OSDs
+
+    Returns:
+        bool: True if osd data consumption difference is <= 2% else False
```
Comment (vasukulkarni) [inline ocs_ci/ocs/cluster.py]: Nice!, we can also take the percent as param if that helps later on, but for now looks good.

**Comment by vasukulkarni** [comment]
Please mark the comments as resolved, so it can be merged.

**Pull request merged**